### DISCUSSION TITLE
Support for Yeelight YLAI003 in Tasmota32 BLE

### DISCUSSION THREAD

**Discussion by taylormia** [arendst/Tasmota#16415] [Ideas]
I noted that the [Yeelight Remote Control](https://tasmota.github.io/docs/Bluetooth_ESP32/#supported-devices) is a BLE sensor supported with a single press and long press on Tasmota32 BLE.

I'm requesting support for the [Yeelight Switch/Button YLAI003](https://a.aliexpress.com/_BS9eKN) which has single press, double press and long press.

When I use the Yeelink YLAI003 with tasmota32c3-bluetooth.bin version 12.1.0.1, I can see the MQTT tele messages send a Btn 0 for single press, Btn 0 for double press and Btn 6 for long press. It seems that the Yeelight protocol is being interpreted correctly by tasmota32 for single and long presses but not for the double press.

There was some [previous discussion](https://github.com/arendst/Tasmota/issues/9926) on this switch/button with additional info.
Can support for double press on the YLAI003 be added?
Thanks.

**Comment by sfromis** [comment]
It would probably be useful if you posted the log you see when trying it out. (I'm not the one to look further into this).

**Comment by taylormia** [comment]
**Tasmota32 Bluetooth ( tasmota32c3-bluetooth.bin) ( version 12.1.1.1)**
Here's the log with Weblog 4 turned on, for single press, double pless and long press,:

**Single Press:**
`17:34:56.504 MQT: tele/tasmota_ble/Unknown9e3274 = {"Time":"2022-09-02T17:34:56","mac":"b460ed9e3274","Btn":0,"RSSI":-45}`

**Double Press:**
`17:34:59.661 MQT: tele/tasmota_ble/Unknown9e3274 = {"Time":"2022-09-02T17:34:59","mac":"b460ed9e3274","Btn":0,"RSSI":-47}`

**Long Press:**
`17:35:03.372 MQT: tele/tasmota_ble/Unknown9e3274 = {"Time":"2022-09-02T17:35:03","mac":"b460ed9e3274","Btn":6,"RSSI":-60}`

I then turned on BLEDebug and got the additional logs below. I notice a change in the last digit of the Decrypted string and payload - with a 0 for Single Press, 1 for Double Press and 2 for Long Press.

**Single Press:**
`17:51:38.941 M32: MI:B460ED9E3274 svc[0] UUID (fe95)`
`17:51:38.943 M32: MI:5859BF072674329EED60B418D976A3EBEF0000004862221F`
`17:51:38.944 M32: Search key for MAC: b460ed9e3274`
`17:51:38.946 M32: Decryption Key found`
`17:51:38.948 M32: Err:1, Decrypted : 10 03 00 00 00`
`17:51:38.949 M32 B460ED9E3274: Payload decrypted`
`17:51:38.951 M32: MI32: slot: 2/7 - ign repeat`
`17:51:38.952 M32: Frame 38, last 38`
`17:51:38.953 M32: MI32: slot: 2/7`
`17:51:38.954 M32: MI2 payload 011003000000`
`17:51:38.960 MQT: tele/tasmota_ble/Unknown9e3274 = {"Time":"2022-09-02T17:51:38","mac":"b460ed9e3274","Btn":0,"Battery":100,"RSSI":-46}`

**Double Press:**
`17:54:53.069 M32: MI:B460ED9E3274 svc[0] UUID (fe95)`
`17:54:53.070 M32: MI:5859BF072874329EED60B4023DC4FF84F5000000E59073F9`
`17:54:53.072 M32: Search key for MAC: b460ed9e3274`
`17:54:53.073 M32: Decryption Key found`
`17:54:53.075 M32: Err:1, Decrypted : 10 03 00 00 01`
`17:54:53.076 M32 B460ED9E3274: Payload decrypted`
`17:54:53.077 M32: MI32: slot: 2/8 - ign repeat`
`17:54:53.079 M32: Frame 40, last 40`
`17:54:53.080 M32: MI32: slot: 2/8`
`17:54:53.082 M32: MI2 payload 011003000001`
`17:54:53.101 MQT: tele/tasmota_ble/Unknown9e3274 = {"Time":"2022-09-02T17:54:53","mac":"b460ed9e3274","Btn":0,"Battery":100,"RSSI":-44}`

**Long Press:**
`17:58:09.325 M32: MI:B460ED9E3274 svc[0] UUID (fe95)`
`17:58:09.327 M32: MI:5859BF072C74329EED60B4B9BE2C26A687000000B3627106`
`17:58:09.328 M32: Search key for MAC: b460ed9e3274`
`17:58:09.330 M32: Decryption Key found`
`17:58:09.331 M32: Err:1, Decrypted : 10 03 00 00 02`
`17:58:09.333 M32 B460ED9E3274: Payload decrypted`
`17:58:09.334 M32: MI32: slot: 2/9 - ign repeat`
`17:58:09.336 M32: Frame 44, last 44`
`17:58:09.337 M32: MI32: slot: 2/9`
`17:58:09.338 M32: MI2 payload 011003000002`
`17:58:09.346 MQT: tele/tasmota_ble/Unknown9e3274 = {"Time":"2022-09-02T17:58:09","mac":"b460ed9e3274","Btn":6,"Battery":100,"RSSI":-71}`

**Comment by barbudor** [comment]
Can't find out by just reading the code
I've ordered one but won't be delivered before 3 weeks

**Comment by taylormia** [comment]
Ok. Thank you.

**Reply by barbudor** [reply]
Just received my Yeelight button
I will look into it this week-end

How do you get the decryption key for it ?

**Comment by taylormia** [comment]
I used Step 1 in the instuctions here: https://custom-components.github.io/ble_monitor/faq#encryption-keys.
In summary:
1. Add the Yeelight button/switch to the MiHome app. You have to specify China Mainland as the Region under Settings in the app to get the Yeelight Wireless Switch YLAI003 option. You should find it in the Power Switch > Remote Control devices section.
2. Use the Xiaomi Cloud Token Extractor to get the keys.

Let me know if you have questions.
Thanks.

**Comment by taylormia** [comment]
Hi,
Any progress on this?

Thanks.

**Reply by barbudor** [reply]
Sorry, I haven't started to look on getting the key
I think I've read somewhere about an python script to get them without installing those chinese app that I wouldn't like to
Does that ring a bell ?

**Reply by sfromis** [reply]
What I've done for BLE pairing with key has been to use a web browser on an Android device, and just click on the link Tasmota provided on the front page.

**Reply by barbudor** [reply]
AFAIK the web page (pvvx or original attc1441) only works with Mija sensors

For the other link (PiotrMachowski/Xiaomi-cloud-tokens-extractor) it looks like this one also requires to first have the App in order to register the BLE device with the cloud 😢

**Comment by taylormia** [comment]
Yes, the MiHome app is required for the token extractor to work. Also, you can delete the app after extracting the token...but the app is required to add the device and get the token from the token extractor script.

**Comment by taylormia** [comment]
Any update on this? Were you able to get the key?

**Reply by barbudor** [reply]
For now I didn't suceed in getting a decryption key using other tools I used
considering that I will NOT install the chinese app on my mobile phone
I tried https://github.com/dnandha/miauth but it doesn't work for now I created an Issue

**Comment by taylormia** [comment]
OK. Thanks
